### DISCUSSION TITLE
Type Compatibility with Apollo Full Stack Example

### DISCUSSION THREAD

**Discussion by mikecfisher** [dotansimha/graphql-code-generator#4992] [General]
Until recently I've yet to see a good example of unit testing resolvers in a TypeScript Apollo Server app. 

I recently discovered Apollo's example repo where they have some nice patterns for unit, integration and e2e testing Apollo Server apps. [Apollo Test Example](https://github.com/apollographql/fullstack-tutorial/blob/master/final/server/src/__tests__/resolvers.query.js)

However in my application while the testing patterns work, it tends to throw many typescript errors. 

Error 1: 

```
This expression is not callable.
  Not all constituents of type 'ResolverFn<Maybe<Maybe<ResolverTypeWrapper<ProductResults>>[]>, {}, Context, RequireFields<QueryProductRoundUpArgs, "category">> | LegacyStitchingResolver<...> | NewStitchingResolver<...>' are callable.
```

This was fixed by turning setting `noSchemaStitching: false` thanks @dotansimha for this. Saw this in another issue. 

Error 2:
`Cannot invoke an object which is possibly 'undefined'.ts`
This is because the resolver type is returned as optional. I can "fix" this by turning: `avoidOptionals: true` but then I get more TypeScript errors in my app due to all resolvers being required. 

Error 3: 
```
Argument of type '{ dataSources: { Product: { find: jest.Mock<any, any>; }; };}' is not assignable to parameter of type 'Context'.
  Type '{ dataSources: { Product: { find: Mock<any, any>; }; }' is missing the following properties from type 'Context': domains, models, logger, login, and 11 more.ts(2345)
```

This is because I have a large context object, however the resolver only requires this one Product datasource. 
I'm creating a mock datasource like this: 

```ts
 const mockContext = {
    dataSources: {
      Product: {find: jest.fn()},
    },
  };
```

But because I'm not passing in the full context I get this TS error. 

If anyone has any ideas on how to fix these issues that would be great. Because the patterns in Apollo's example repo are great and feel like they should be the canonical way to test Apollo Server apps. However its tough to do that with the way graphql-codegen generates it's types for me.

**Comment by dotansimha** [comment]
@mikefisher84 you can use different values for `avoidOptionals`  in each plugin. So you can change the generates types signature, without applying that to resolvers plugins.
Can you please share how your codegen.yml looks like?
